pinot-tools/src/main/java/com/linkedin/pinot/tools/admin/command/MergeConfigsCommand.java

```diff
@@ -0,0 +1,689 @@
+  // Object, but makes the types below clearer)
+  private <TYPE> boolean executeInternal() throws Exception {
+    // Build a list of all input files and their associated configuration profiles
+    LOGGER.info("Searching for configs...");
+
+    File inputDir = new File(_inputDir);
+    if (!inputDir.exists()) {
+      throw new RuntimeException("Input directory " + inputDir + " does not exist!");
+    } else if (!inputDir.isDirectory()) {
+      throw new RuntimeException("Input directory " + inputDir + " is not a directory!");
+    }
+
+    List<File> configFiles = List.ofAll(FileUtils.listFiles(inputDir, CONFIG_FILE_EXTENSIONS, true));
+
+    List<Tuple2<File, Set<String>>> configFilesAndProfiles = configFiles.map(configFile -> {
+      File currentFile = configFile.getParentFile();
+      Set<String> configurationProfiles = HashSet.empty();
+
+      // Iterate through parent files until the config directory is found to build the set of configuration profiles
+      // that apply to this particular configuration file
+      while(!currentFile.equals(inputDir)) {
+        configurationProfiles = configurationProfiles.add(currentFile.getName());
+
+        File parentOfCurrentFile = currentFile.getParentFile();
+        if (parentOfCurrentFile != null) {
+          currentFile = parentOfCurrentFile;
+        } else {
+          throw new RuntimeException("Failed to get parent of " + currentFile.getPath() + " while finding configuration profiles for " + configFile.getPath());
+        }
+      }
+
+      return Tuple.of(configFile, configurationProfiles);
+    });
+
+    LOGGER.info("Found {} config files", configFilesAndProfiles.length());
+
+    // Load all the configuration profiles
+    LOGGER.info("Loading all configuration profiles...");
+
+    if (_profileDir == null) {
+      _profileDir = "." + File.separator + "profiles";
+    }
+    File profileDir = new File(_profileDir);
+
+    List<Tuple2<String, Config>> profileConfigurations;
+    if (!profileDir.exists()) {
+      System.out.println("Input directory " + inputDir + " does not exist.");
+      profileConfigurations = List.empty();
+    } else if (!inputDir.isDirectory()) {
+      throw new RuntimeException("Input directory " + inputDir + " is not a directory!");
+    } else {
+      profileConfigurations = List
+          .ofAll(FileUtils.listFiles(profileDir, CONFIG_FILE_EXTENSIONS, true))
+          .map(file -> {
+            Config config = loadConfigFromFile(file);
+            String configName = file.getName().replaceAll("\\.conf$", "");
+
+            return Tuple.of(configName, config);
+          });
+    }
+
+    LOGGER.info("Loaded {} configuration profiles: {}", profileConfigurations.length(), profileConfigurations.map(Tuple2::_1).asJava());
+
+    // Build a map of value to configuration key for each configuration profile
+    Map<TYPE, Map<String, Set<String>>> valueToProfileAndKeys = profileConfigurations
```

Comment:
this is really complex to comprehend. It almost starting to look like scala/functional programming. Let's avoid having so many anonymous functions, makes it really hard to debug due to complex stacktrace. Please break this down or convert it into traditional java loop.

Reply:
Converting it to loop-based code only makes it longer, not necessarily clearer. It's a pretty long (and tedious) task. I'll see what my calendar looks like for scheduling this and if the readability improvements warrant the change.